probe_src/probe_py/analysis.py

```diff
@@ -1,53 +1,64 @@
 import typing
-import networkx
-from .parse_probe_log import ProvLog, Op, CloneOp, ExecOp, WaitOp, CLONE_THREAD
+import networkx as nx
+from .parse_probe_log import ProcessTreeProvLog, Op, CloneOp, ExecOp, WaitOp, CLONE_THREAD
 
-
-def construct_process_graph(process_tree_prov_log: ProvLog) -> str:
+def construct_process_graph(process_tree_prov_log: ProcessTreeProvLog):
     """
     Construct a happens-before graph from process_tree_prov_log
 
     The graph shows the order that prov ops happen.
     """
-    Node: typing.TypeAlias = tuple[int, int, int, Op | str]
+    # [pid, exec_epoch_no, tid, Op | str, node_count, op_index, _time]
+    Node: typing.TypeAlias = tuple[int, int, int, Op | str, int, int, int]
     nodes = list[Node]()
     program_order_edges = list[tuple[Node, Node]]()
     proc_to_ops = dict[tuple[int, int, int], list[Node]]()
     last_exec_epoch = dict[int, int]()
-
-    for pid, process in process_tree_prov_log.processes.items():
+    global node_count
+    node_count = int(1)
+    for (pid, _time), process in process_tree_prov_log.processes.items():
         for exec_epoch_no, exec_epoch in process.exec_epochs.items():
+            # to find the last executing epoch of the process
             last_exec_epoch[pid] = max(last_exec_epoch.get(pid, 0), exec_epoch_no)
 
             # Reduce each thread to the ops we actually care about
             for tid, thread in exec_epoch.threads.items():
                 context = (pid, exec_epoch_no, tid)
-
```

Comment:
In my opinion, `list[Node]()` is more readable.

Reply:
Fixed this!